Proper docs site (#59)
dabapps/django-readers: docs/tutorial.md

## Installing django-readers

Install from [PyPI](https://pypi.org/project/django-readers/):

```shell
pip install django-readers

> ```

Reviewer: Would this be better suited in the introduction?

## Example models

For the purposes of this tutorial, we're going to use a set of models representing **books**, **authors** and **publishers**. A book has exactly one publisher, but can have many authors. Here are the Django models we're using:

```python
class Publisher(models.Model):
    name = models.CharField(max_length=100)
    address = models.TextField()

class Author(models.Model):
    name = models.CharField(max_length=100)
    email = models.EmailField()

class Book(models.Model):
    title = models.CharField(max_length=100)
    authors = models.ManyToManyField(Author)
    publisher = models.ForeignKey(Publisher)
    publication_date = models.DateField()
```

We're going to go through a step-by-step set of requirements from an imaginary client who has asked us to build a simple book listing application. We're going to assume the Django project itself has already been set up, and focus on a single view and a single template.

## Requirement 1: a list of book titles (without `django-readers`)

> First, the client wants us to build a page containing a list of all of the books in the database, showing only their titles. The standard Django approach to this would be as follows:

Reviewer: ```suggestion
First, we want to build a page containing a list of all of the books in the database, showing only their titles. The standard Django approach to this would be as follows:
```

I'm thinking not everyone reading these docs will be working with clients so might read a bit strangely to them.

If you agree with this change, don't forget to change it in other places too.

Reviewer 2: i actually prefer using the pronoun "you" as I am closer to the Google styleguide as mentioned here in https://ddbeck.com/first-person-plural-is-questionable/

Author: Agreed about "you" vs "we", I'll tweak this. I think using an imaginary client as the motivation for each step in the tutorial is probably OK, a lot of people working with Django do have clients, and those that don't can substitute a product manager (or whoever) making the requests.

!!! note
    YAGNI, "You Aren't Gonna Need It", it's a [well-understood mantra](https://martinfowler.com/bliki/Yagni.html) in software development. It means that you should only make changes to your software (including adding abstraction and generalising code structures) when you are sure that you absolutely need them, and not before. Related to this is the famous quote from Donald Knuth: "premature optimization is the root of all evil". This suggests that usually performance concerns fall under YAGNI: you shouldn't spend time making code fast until its slowness becomes a real problem for users.

    As a counterpoint to this advice, Simon Willison coined the phrase [PAGNI](https://simonwillison.net/2021/Jul/1/pagnis/), "Probably Are Gonna Need It". PAGNI applies in situations "when the cost of adding something later is so dramatically expensive compared with the cost of adding it early on that it’s worth taking the risk [...] when you know from experience that an initial investment will pay off many times over."

>     `django-readers` (and its sister project `django-zen-queries`) were built partly as a result of a strong belief (gained through many years of experience) that understanding and controlling your application's database querying behaviour firmly falls into the PAGNI category. This is particularly true of applications that you know are going to be complex: any code abstraction patterns you decide to follow in order to manage the complexity _must_ take into account query patterns or they are highly likely to cause terrible performance problems. This is the heart of the [object-relational impedance mismatch](https://en.wikipedia.org/wiki/Object%E2%80%93relational_impedance_mismatch).

Reviewer: This bit is lovely and quite an important part of the motivation behind `django-readers`

I feel like it needs a more prominent place, perhaps in the `Explanation`.

Though it does read nicely here as well.

Reviewer 2: @j4mie I strongly agree with @nvlaarhoven about the following

> I feel like it needs a more prominent place, perhaps in the Explanation.

Which prominent place is up to you. BUt yes, this should be somewhere more prominent.

Reviewer 2: Thanks for mentioning `zen-queries`, @nvlaarhoven . I had starred the repo but completely forgot about it until you pointed this out again. Thank you. 🙇🏻‍♂️

Author: I'll move this to the Explanation page.

```python
from django_readers import specs

def book_list(request):
    spec = [
        "title",
    ]
    prepare, project = specs.process(spec)
    queryset = prepare(Book.objects.all())
    return render(
        "book_list.html",
        {"books": [project(book) for book in queryset]},
    )
```

As you can see, at its simplest a `django-readers` spec is just a list of field names. `specs.process` takes the spec and returns _a pair of functions_, which (by convention) we call `prepare` and `project`.

!!! note
    `prepare` and `project` are `django-readers`-specific terms. There is one other `django-readers` concept beginning with P, `produce`, which will be introduced fully later, but here are their rough definitions to set the scene:

    A `prepare` function takes a queryset and returns another queryset with one or more modifications applied (like `prefetch_related` or `only`: just like chaining queryset methods). This process is called "preparing the queryset".

    A `produce` function (a "producer") takes a model instance and returns ("produces") a value derived from that instance. Often this is the value of an attribute on the model.

>     A `project` function (a "projector") takes a model instance and returns a dictionary (the "projection") which maps one or more names (the dictionary keys) to one or more values derived from the instance (often the values of one or more attributes and possibly nested dictionaries representing related models). It _projects_ your data layer into your application's business logic domain. Think of the dictionary returned by a projector (the "projection") as the simplest possible domain object. Generally speaking, it's not necessary to write your own projector functions: you can wrap a producer function to bind the value it returns to a name.

Reviewer 2: Assuming the diataxis quadrant and this is for tutorial/quickstart, i think this entire page should start with the following

## Retrieve 1 instance

in views.py
```python

from django_readers import specs

def book_details(request, pk):
    spec = [
        "title",
    ]
    prepare, project = specs.process(spec)
    queryset = prepare(Book.objects.all()).get(pk=pk)
	book = project(queryset)
    return render(
        "book_details.html",
        {"book": book},
    )
```

## Retrieve a list of many instances

this is in your intro but no harm repeating

>     A `project` function (a "projector") takes a model instance and returns a dictionary (the "projection") which maps one or more names (the dictionary keys) to one or more values derived from the instance (often the values of one or more attributes and possibly nested dictionaries representing related models). It _projects_ your data layer into your application's business logic domain. Think of the dictionary returned by a projector (the "projection") as the simplest possible domain object. Generally speaking, it's not necessary to write your own projector functions: you can wrap a producer function to bind the value it returns to a name.

Author: I think we've never hit that gotcha because we don't tend to use `__str__` much. Usually `__str__` would just return the value of a field on the model (`self.title`, say), so I think it's clearer to just explicitly include that field in your spec. If you do have some more complex logic in `__str__` then your `display` workaround is fine, but I'm not sure that needs to be right there in the tutorial - I think (hope!) that people can work that out for themselves once they've read the rest of the docs.

Regarding putting the examples at the top of the tutorial: doesn't that rather derail the narrative? I think building up to those examples guides understanding more clearly.
